src/embind/embind.js

```diff
@@ -729,39 +729,79 @@ var LibraryEmbind = {
+        decodeString = UTF16ToString;
+        encodeString = stringToUTF16;
+        lengthBytesUTF = lengthBytesUTF16;
         getHeap = function() { return HEAPU16; };
         shift = 1;
     } else if (charSize === 4) {
+        decodeString = UTF32ToString;
+        encodeString = stringToUTF32;
+        lengthBytesUTF = lengthBytesUTF32
         getHeap = function() { return HEAPU32; };
         shift = 2;
     }
     registerType(rawType, {
         name: name,
         'fromWireType': function(value) {
-            var HEAP = getHeap();
             var length = HEAPU32[value >> 2];
-            var a = new Array(length);
-            var start = (value + 4) >> shift;
-            for (var i = 0; i < length; ++i) {
-                a[i] = String.fromCharCode(HEAP[start + i]);
+            var HEAP = getHeap();
+            var str;
+            //ensure null termination at one-past-end byte if not present yet
+            var endChar = HEAP[(value + 4 + length * charSize) >> shift];
+            var endCharSwap = 0;
+            if(endChar != 0)
+            {
+                endCharSwap = endChar;
+                HEAP[(value + 4 + length * charSize) >> shift] = 0;
             }
+
+            var decodeStartPtr = value + 4;
+            //looping here to support possible embedded '0' bytes
+            for (var i = 0; i <= length; ++i) {
+                var currentBytePtr = value + 4 + i * charSize;
+                if(HEAP[currentBytePtr >> shift] == 0)
+                {
+                    var stringSegment = decodeString(decodeStartPtr);
+                    if(str === undefined)
+                        str = stringSegment;
+                    else
+                    {
+                        str += String.fromCharCode(0);
+                        str += stringSegment;
+                    }
+                    decodeStartPtr = currentBytePtr + charSize;
+                }
+            }
+
+            if(endCharSwap != 0)
+                HEAP[(value + 4 + length * charSize) >> shift] = endCharSwap;
+
             _free(value);
-            return a.join('');
+
+            return str;
         },
         'toWireType': function(destructors, value) {
-            // assumes 4-byte alignment
-            var HEAP = getHeap();
-            var length = value.length;
-            var ptr = _malloc(4 + length * charSize);
-            HEAPU32[ptr >> 2] = length;
-            var start = (ptr + 4) >> shift;
-            for (var i = 0; i < length; ++i) {
-                HEAP[start + i] = value.charCodeAt(i);
+            if (value instanceof ArrayBuffer) {
+                value = new Uint8Array(value);
+            }
+
+            var valueIsOfTypeString = (typeof value === 'string');
+
+            if (!(valueIsOfTypeString || value instanceof Uint8Array || value instanceof Uint8ClampedArray || value instanceof Int8Array)) {
+                throwBindingError('Cannot pass non-string to std::basic_string');
```

Comment:
This should say `std::wstring`?

Reply:
Well, since it could be `std::wstring`, `std::u16string` or `std::u32string` I stayed general by saying `std::basic_string`

Reply:
Would that:

> Cannot pass non-string to `std::wstring`, `std::u16string` or `std::u32string`

Be better?

Reply:
To *me specifically* that sounds better, but I didn't know that basic_string was the base of wstring et al, so it looked like a copy-paste error. https://en.cppreference.com/w/cpp/string/basic_string lists all those, so maybe basic_string is fine.

Having said that, I imagine this code breaks if you pass any basic_string that isn't wstring u16string or u32string, so probably better to be explicit about what we support.

Reply:
@jgravelle-google I just realized that the name is sent as parameter to the function. Now it's going to print the right type name everytime, even when merging the other patch.